models/index.js

```diff
@@ -54,14 +54,52 @@ exports.fetchUsers = () => {
   });
 };
 
-exports.fetchAllArticles = () => {
-  return db
-    .query(
-      "SELECT articles.*, COUNT(comments.article_id)::int AS comment_count FROM articles LEFT JOIN comments ON comments.article_id = articles.article_id GROUP BY articles.article_id, comments.article_id ORDER BY created_at DESC;"
-    )
-    .then(({ rows }) => {
-      return rows;
-    });
+exports.fetchAllArticles = (sort_by = "created_at", order = "desc", topic) => {
+  const validSortOptions = [
+    "author",
+    "title",
+    "article_id",
+    "topic",
+    "created_at",
+    "votes",
+    "comment_count",
+  ];
+  const validOrderOptions = ["asc", "desc"];
+  if (
+    !validSortOptions.includes(sort_by) ||
+    !validOrderOptions.includes(order)
+  ) {
+    return Promise.reject({ status: 400, message: "400 - Invalid query" });
+  }
+
+  const queryValues = [];
+  let queryStr = `SELECT articles.*, COUNT(comments.article_id)::int AS comment_count FROM articles LEFT JOIN comments ON comments.article_id = articles.article_id`;
+
+  if (topic) {
+    queryValues.push(topic);
+    queryStr += " WHERE articles.topic=$1";
+  }
+
+  queryStr += ` GROUP BY articles.article_id, comments.article_id ORDER BY ${sort_by} ${order};`;
+
+  if (topic) {
+    return db
```

Comment:
could this not all be done inside the one conditional statement for if (topic)?

Reply:
I think the WHERE articles.topic on line 80 needs to be there to insert at the correct location in the queryStr - then the If(topic) on line 85 catches there being a topic then checks if the topic exists. Line 83 needs to be there for the correct ordering in the queryStr again. That make sense?